blacklists.py

```diff
@@ -140,17 +147,174 @@ def exists(self, item: Union[str, dict]):
                 splat = x.split('\t')
                 if len(splat) == 3 and splat[2].strip() == item:
                     return True, i
+        return False, -1
+
+
+class YAMLParserCIDR(BlacklistParser):
+    """
+    YAML parser for IP blacklist (name suggests we should move to proper CIDR eventually).
+
+    Base class for parsers for YAML files with simple schema validation.
+    """
+    # Remember to update the schema version if any of this needs to be changed
+    SCHEMA_VERSION = '2019120601'  # yyyy mm dd id
+    SCHEMA_VARIANT = 'yaml_cidr'
+    SCHEMA_PRIKEY = 'ip'
+
+    def __init__(self, filename):
+        super().__init__(filename)
+
+    def _parse(self, keep_disabled=False):
+        with open(self._filename, 'r', encoding='utf-8') as f:
+            y = yaml.safe_load(f)
+        assert y['Schema'] == self.__class__.SCHEMA_VARIANT
+        assert y['Schema_version'] <= self.__class__.SCHEMA_VERSION
+        for item in y['items']:
+            if not keep_disabled and 'disable' in item and item['disable']:
+                continue
+            yield item
+
+    def parse(self):
+        for item in self._parse():
+            yield item[self.__class__.SCHEMA_PRIKEY]
+
+    def _write(self, callback):
+        d = {
+            'Schema': self.__class__.SCHEMA_VARIANT,
+            'Schema_version': self.__class__.SCHEMA_VERSION,
+            'items': sorted(
+                self._parse(keep_disabled=True),
+                key=lambda x: x[self.__class__.SCHEMA_PRIKEY])
+        }
+        callback(d)
+        with open(self._filename, 'w', encoding='utf-8') as f:
+            yaml.dump(d, f)
+
+    def _validate(self, item):
+        ip_regex = regex.compile(
+            r'(?(DEFINE)(?P<octet>'
+            '' r'1[0-9]{0,2}|2(?:[0-4][0-9])??|25[0-5]?|2[6-9]|[3-9][0-9]?))'
+            r'^(?&octet)(?:\.(?&octet)){3}$'
+        )
+        if 'ip' in item:
+            if not ip_regex.match(item['ip']):
+                raise ValueError('Field "ip" is not a valid IP address: {0}'.format(
+                    item['ip']))
+            '''
+            if 'cidr' in item:
```

Comment:
What about unioning `ip` and `cidr` so that a CIDR range can appear as the value to an `ip` key, while treating pure IPs as if they had a `/32` suffix. Many places I see supports mixing them.

Reply:
I plan on something like that going forward, but for the time being this simply moves the data we have now. My proposal would be to use `cidr: '12.34.56.0/24'` but allow 'ip: '12.34.56.78'` as a convenience shorthand for `cidr: '12.34.56.78/32'`

Reply:
Sorry if I didn't made myself clear. I mean we accept both `ip: 1.1.1.1` and `ip: 1.1.1.1/24` forms and discard the `cidr` key, just as you'd write in a firewall rule (you don't put raw IPs and CIDR blocks into two inputs, right?).

Reply:
Hmm, but a CIDR isn't an IP. But that's for the future anyway.